### ISSUE TITLE
no-this-alias is a bit noisy

### ISSUE THREAD

**Issue by callionica** [denoland/deno_lint#360]
It's common to want to assign `this` to a variable when implementing tree methods and there's no harm in it (as below). Shouldn't this rule only fire if the alias is used in a local function or escapes the current method?

```
get root(): this {
        // deno-lint-ignore no-this-alias
        let primary = this;
        while (true) {
            if (primary.parent !== undefined) {
                primary = primary.parent;
            } else {
                return primary;
            }
        }
    }

```

**Comment by bartlomieju** [comment]
CC @magurotuna what do you think?

**Comment by magurotuna** [comment]
Yes, indeed there's no harm with the `this` alias in the snippet. But in this particular case, there's a way to achieve the same goal without storing `this` in a variable. I mean, if I were to rewrite the logic, I would do:

```ts
get root(): this {
  if (this.parent === undefined) {
    return this;
  }
  return this.parent.root;
}
```

I think this looks much clearer and simpler.

If there are cases where we definitely need to store `this` in a variable to do something, then I think we should fix the logic of `no-this-alias` rule, but I don't come up with those cases right now.

**Comment by waptik** [comment]
> Yes, indeed there's no harm with the `this` alias in the snippet. But in this particular case, there's a way to achieve the same goal without storing `this` in a variable. I mean, if I were to rewrite the logic, I would do:
> 
> ```ts
> get root(): this {
>   if (this.parent === undefined) {
>     return this;
>   }
>   return this.parent.root;
> }
> ```
> 
> I think this looks much clearer and simpler.
> 
> If there are cases where we definitely need to store `this` in a variable to do something, then I think we should fix the logic of `no-this-alias` rule, but I don't come up with those cases right now.

Any updates? I have a class and in a method where i use cheerio to grab some data, i have a property say `name`.
And to have access to the right value for `this.name` within a cheerio element map (`$(element).map(function(){ ... })`) inside a method, i have to first assign `this` to a variable(eg: `const parent = this`) and inside the loop, i can have access to the value by using `parent.name`. If i don't do that, i get the wrong value for `this.name`.

I have disabled the warning.

**Comment by Nikaple** [comment]
@waptik You should use arrow functions. See: https://typescript-eslint.io/rules/no-this-alias/

**Comment by alan-khosro** [comment]
An example that sometimes assigning `this` is necessary:

```ts
class Router {
        routs: Route[] = []
        
        push() {
        // do the push
        }
        
        // decorator
	assign = (
		handler: Function,
		context: { addInitializer: AddInitializer }
	) => {
		// rename `this` to router, because inside initializer `this` is the caller class which we need it
		// deno-lint-ignore no-this-alias
		const router = this; // here this means Router class
		context.addInitializer(function () {
			router.push(handler.bind(this)); // this here means caller class
		});
	};
}
```
I wrote all functions as arrow functions to avoid the discussions "use arrow function".

If you have a way to avoid this assignment, please let me know. But, as I understood, sometimes it is necessary to work with two `this` and you need to assign outer `this` to some variable.
